params/config.go

```diff
@@ -910,6 +910,11 @@ func (c *ChainConfig) Rules(num *big.Int, isMerge bool, timestamp uint64) Rules
 	if chainID == nil {
 		chainID = new(big.Int)
 	}
+	if c.IsShanghai(num, timestamp) {
```

Comment:
I think it's pretty weird. It can happen that:

London, Merge, Shanghai forks are all configured, but merge is not activated (ttd is not reached). In this case even the London is activated and pre-configured shanghai timestamp is reached, Shanghai shouldn't be enabled.